File: app/src/main/java/com/bike/computer/GpxRoute.java
package com.bike.computer;

import java.io.IOException;
import java.util.ArrayList;
import java.util.List;
import java.util.Set;
import java.util.concurrent.TimeUnit;
import kotlin.collections.CollectionsKt;
import kotlin.collections.SetsKt;
import kotlin.io.CloseableKt;
import kotlin.jvm.internal.Intrinsics;
import kotlin.text.MatchResult;
import kotlin.text.Regex;
import kotlin.text.RegexOption;
import kotlin.text.StringsKt;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.Response;
import okhttp3.ResponseBody;

public final class GpxRoute {
    public static final GpxRoute INSTANCE = new GpxRoute();
    private static final OkHttpClient http = new OkHttpClient.Builder().connectTimeout(20, TimeUnit.SECONDS).readTimeout(30, TimeUnit.SECONDS).build();
    private static final Regex ptRe = new Regex("<(?:trkpt|rtept)\\b([^>]*)>", RegexOption.IGNORE_CASE);
    private static final Regex latRe = new Regex("lat=\"([-0-9.]+)\"", RegexOption.IGNORE_CASE);
    private static final Regex lonRe = new Regex("lon=\"([-0-9.]+)\"", RegexOption.IGNORE_CASE);
    private static final Regex nameRe = new Regex("<name>(.*?)</name>", SetsKt.setOf(new RegexOption[]{RegexOption.IGNORE_CASE, RegexOption.DOT_MATCHES_ALL}));

    private GpxRoute() {
    }

    public final String download(String url) throws IOException {
        String body;
        Intrinsics.checkNotNullParameter(url, "url");
        String u = StringsKt.startsWith(url, "http", false) ? url : "https://" + url;
        Response responseExecute = http.newCall(new Request.Builder().url(u).header("User-Agent", "BikeComputer").build()).execute();
        try {
            Response r = responseExecute;
            ResponseBody responseBodyBody = r.body();
            if (responseBodyBody == null || (body = responseBodyBody.string()) == null) {
                body = "";
            }
            if (!r.isSuccessful()) {
                throw new RuntimeException("HTTP " + r.code());
            }
            CloseableKt.closeFinally(responseExecute, null);
            return body;
        } finally {
        }
    }

    public final String name(String gpx) {
        List<String> groupValues;
        String str;
        String it;
        Intrinsics.checkNotNullParameter(gpx, "gpx");
        MatchResult matchResultFind$default = nameRe.find(gpx, 0);
        if (matchResultFind$default == null || (groupValues = matchResultFind$default.getGroupValues()) == null || (str = groupValues.get(1)) == null || (it = StringsKt.trim((CharSequence) str).toString()) == null) {
            return null;
        }
        if (it.length() > 0) {
            return it;
        }
        return null;
    }

    public final List<double[]> parse(String gpx) {
        List<String> groupValues;
        String str;
        Double doubleOrNull;
        List<String> groupValues2;
        String str2;
        Double doubleOrNull2;
        Intrinsics.checkNotNullParameter(gpx, "gpx");
        ArrayList pts = new ArrayList();
        for (MatchResult m : kotlin.sequences.SequencesKt.asIterable(ptRe.findAll(gpx, 0))) {
            String a = m.getGroupValues().get(1);
            MatchResult matchResultFind$default = latRe.find(a, 0);
            if (matchResultFind$default != null && (groupValues = matchResultFind$default.getGroupValues()) != null && (str = groupValues.get(1)) != null && (doubleOrNull = StringsKt.toDoubleOrNull(str)) != null) {
                double lat = doubleOrNull.doubleValue();
                MatchResult matchResultFind$default2 = lonRe.find(a, 0);
                if (matchResultFind$default2 != null && (groupValues2 = matchResultFind$default2.getGroupValues()) != null && (str2 = groupValues2.get(1)) != null && (doubleOrNull2 = StringsKt.toDoubleOrNull(str2)) != null) {
                    double lon = doubleOrNull2.doubleValue();
                    pts.add(new double[]{lon, lat});
                }
            }
        }
        return pts;
    }

    public static /* synthetic */ List toWaypoints$default(GpxRoute gpxRoute, List list, double d, int i, int i2, Object obj) {
        if ((i2 & 2) != 0) {
            d = 1200.0d;
        }
        if ((i2 & 4) != 0) {
            i = 24;
        }
        return gpxRoute.toWaypoints(list, d, i);
    }

    public final List<double[]> toWaypoints(List<double[]> points, double spacingM, int maxVias) {
        Intrinsics.checkNotNullParameter(points, "points");
        if (points.size() <= 2) {
            return points;
        }
        double total = length(points);
        double spacing = Math.max(spacingM, total / ((double) maxVias));
        ArrayList out = new ArrayList();
        out.add(CollectionsKt.first((List) points));
        double acc = 0.0d;
        int size = points.size() - 1;
        for (int i = 1; i < size; i++) {
            acc += hav(points.get(i - 1), points.get(i));
            if (acc >= spacing) {
                out.add(points.get(i));
                acc = 0.0d;
            }
        }
        out.add(CollectionsKt.last((List) points));
        return out;
    }

    private final double length(List<double[]> p) {
        double s = 0.0d;
        int size = p.size();
        for (int i = 1; i < size; i++) {
            s += hav(p.get(i - 1), p.get(i));
        }
        return s;
    }

    private final double hav(double[] a, double[] b) {
        double p1 = Math.toRadians(a[1]);
        double p2 = Math.toRadians(b[1]);
        double dp = Math.toRadians(b[1] - a[1]);
        double dl = Math.toRadians(b[0] - a[0]);
        double d = 2;
        double it = Math.sin(dp / d);
        double dCos = Math.cos(p1) * Math.cos(p2);
        double it2 = Math.sin(dl / d);
        double h = (it * it) + (dCos * it2 * it2);
        return d * 6371000.0d * Math.asin(Math.sqrt(h));
    }
}

File: app/src/main/java/com/bike/computer/GpxRoute.kt
package com.bike.computer

import okhttp3.OkHttpClient
import okhttp3.Request
import java.io.IOException
import java.util.concurrent.TimeUnit

/** Downloads and parses GPX track/route files into point lists and via waypoints. */
object GpxRoute {
    private val http = OkHttpClient.Builder()
        .connectTimeout(20, TimeUnit.SECONDS)
        .readTimeout(30, TimeUnit.SECONDS)
        .build()
    private val ptRe = Regex("<(?:trkpt|rtept)\\b([^>]*)>", RegexOption.IGNORE_CASE)
    private val latRe = Regex("lat=\"([-0-9.]+)\"", RegexOption.IGNORE_CASE)
    private val lonRe = Regex("lon=\"([-0-9.]+)\"", RegexOption.IGNORE_CASE)
    private val nameRe = Regex(
        "<name>(.*?)</name>",
        setOf(RegexOption.IGNORE_CASE, RegexOption.DOT_MATCHES_ALL),
    )

    @Throws(IOException::class)
    fun download(url: String): String {
        val u = if (url.startsWith("http")) url else "https://$url"
        http.newCall(
            Request.Builder().url(u).header("User-Agent", "BikeComputer").build(),
        ).execute().use { r ->
            val body = r.body?.string() ?: ""
            if (!r.isSuccessful) throw RuntimeException("HTTP ${r.code}")
            return body
        }
    }

    fun name(gpx: String): String? {
        val name = nameRe.find(gpx)?.groupValues?.get(1)?.trim() ?: return null
        return name.ifEmpty { null }
    }

    fun parse(gpx: String): List<DoubleArray> {
        val pts = ArrayList<DoubleArray>()
        for (m in ptRe.findAll(gpx)) {
            val a = m.groupValues[1]
            val lat = latRe.find(a)?.groupValues?.get(1)?.toDoubleOrNull() ?: continue
            val lon = lonRe.find(a)?.groupValues?.get(1)?.toDoubleOrNull() ?: continue
            pts.add(doubleArrayOf(lon, lat))
        }
        return pts
    }

    @JvmOverloads
    fun toWaypoints(
        points: List<DoubleArray>,
        spacingM: Double = 1200.0,
        maxVias: Int = 24,
    ): List<DoubleArray> {
        if (points.size <= 2) return points
        val total = length(points)
        val spacing = maxOf(spacingM, total / maxVias)
        val out = ArrayList<DoubleArray>()
        out.add(points.first())
        var acc = 0.0
        for (i in 1 until points.size - 1) {
            acc += hav(points[i - 1], points[i])
            if (acc >= spacing) {
                out.add(points[i])
                acc = 0.0
            }
        }
        out.add(points.last())
        return out
    }

    private fun length(p: List<DoubleArray>): Double {
        var s = 0.0
        for (i in 1 until p.size) s += hav(p[i - 1], p[i])
        return s
    }

    private fun hav(a: DoubleArray, b: DoubleArray): Double {
        val p1 = Math.toRadians(a[1])
        val p2 = Math.toRadians(b[1])
        val dp = Math.toRadians(b[1] - a[1])
        val dl = Math.toRadians(b[0] - a[0])
        val s1 = Math.sin(dp / 2)
        val c = Math.cos(p1) * Math.cos(p2)
        val s2 = Math.sin(dl / 2)
        val h = s1 * s1 + c * s2 * s2
        return 2 * 6371000.0 * Math.asin(Math.sqrt(h))
    }
}
